openhtmltopdf-core/src/main/java/com/openhtmltopdf/util/XRLog.java

```diff
@@ -66,176 +66,218 @@ private static String registerLoggerByName(final String loggerName) {
      */
     public static List<String> listRegisteredLoggers() {
         // defensive copy
-        return new ArrayList<String>(LOGGER_NAMES);
+        return new ArrayList<>(LOGGER_NAMES);
     }
 
+    public enum LogMessageId {
```

Comment:
I think `LogMessageId` will be big enough to warrant its own file.

Also, I'm worried that since we have a lot of code paths that are not tested, that we will get into trouble with mismatching argument counts for the message format. One way to alleviate this problem would be to have an message enum for each arg count and a corresponding method which accepted that enum and the correct number of args. Then we would make the enums implement a common interface, which is what the user would see. Interested in your feedback.

Reply:
agree about the new file. 

About the mismatching count of parameters/message, I agree we should be able to statically prevent the misuse. I'm wondering if there is an alternative way to implement it, losing the single enum type is unfortunate.

Reply:
done, I think it will work :)